cartographer_grpc/framework/service.cc

```diff
@@ -80,33 +84,42 @@ void Service::HandleNewConnection(Rpc* rpc, bool ok) {
     active_rpcs_.Remove(rpc);
   }
 
-  // TODO(cschuet): Request next read for the new connection.
+  if (ok) {
+    // For request-streaming RPCs ask the client to start sending requests.
+    rpc->RequestStreamingReadIfNeeded();
+  }
 
-  // Create new active rpc to handle next connection.
-  Rpc* next_rpc = active_rpcs_.Add(cartographer::common::make_unique<Rpc>(
-      rpc->method_index(), rpc->server_completion_queue(),
-      rpc->rpc_handler_info(), this));
+  // Create new active rpc to handle next connection and register it for the
+  // incoming connection.
+  active_rpcs_.Add(rpc->Clone())->RequestNextMethodInvocation();
+}
 
-  RequestNextMethodInvocation(rpc->method_index(), next_rpc,
-                              rpc->server_completion_queue());
+void Service::HandleRead(Rpc* rpc, bool ok) {
+  if (ok) {
+    rpc->OnRequest();
+    rpc->RequestStreamingReadIfNeeded();
+    return;
+  }
+
+  // Reads completed.
+  rpc->OnReadsDone();
```

Comment:
Here, "ok" seems to have a different meaning, more like "continue".

Reply:
Somehow. I mean you did ask for a READ in a situation where the client already indicated that reads are done. gRPC indicates that as a failed read and it is AFAIC the only way to learn about client-side read done.